feat-fix(server_openvr): Add timewarped compositing to Windows OpenVR compositor

shinyquagsire23 opened the pull request:
Fixes juddery menus on Riven, and possibly judder on other OpenXR apps (HUDs, 2D menus, text, etc rendered at higher resolutions are the most likely candidates).

[Before](https://youtu.be/cuP-xulPFYY) vs [After](https://youtu.be/ShsK4-j5hZQ)

**tl;dr on compositing**
- To composite with timewarp, we need three things: The exact view poses the layer frame was rendered at, the headset's per-eye view frustum tangents, and a renderer which can render a quad with positions/sizes in meters.
  - FrameRender is now passed the pose of each layer, the view tangents, and the eye transforms to make this happen.
- The layer's quad is centered at 0,0 and extended 1 unit out in each direction, and then each direction is multiplied by the view tangents.
  - Each layer's quad is projected forward by 700m (to "infinity") and scaled proportionally, to remove small positional differences from IPD and the user's head rotation.
  - If the layer is at the exact current view position (which is always the case for layer 0), it will be projected pixel-perfect onto the framebuffer.
  - Layers indices >=1 only ever seem to appear in OpenXR-based titles, I think OpenVR started compositing its app layers into layer 0 on our behalf within the last 6 months (the dashboard used to be juddery as well).
- For simplicity, and to make debugging a bit easier, I did all the vertex math on the CPU with the DirectX math functions (it's only 8 vertices and DirectX has SIMD support anyway).
  - I also split the draw command into two calls with different viewports/scissor rects, since the quads could cross between eyes after the transform
- Also I removed the depth buffer for now, since we were not actually using it. If we ever decide to try and get fancy with depth buffers we can add it back I guess.

zmerp commented on alvr/server_openvr/cpp/platform/win32/FrameRender.cpp:
```diff
@@ -585,39 +691,123 @@ bool FrameRender::RenderFrame(
+                      -1.0f * viewProj[0].vBottomRight.v[1] * depth * m,
+                      -depth,
+                      1.0f } } } };
+#else
+        DirectX::XMVECTORF32 vertsL_VF32[4]
+            = { { { { -1.0f * depth * m, 1.0f * depth * m, -depth, 1.0f } } },
+                { { { 1.0f * depth * m, -1.0f * depth * m, -depth, 1.0f } } },
+                { { { 1.0f * depth * m, 1.0f * depth * m, -depth, 1.0f } } },
+                { { { -1.0f * depth * m, -1.0f * depth * m, -depth, 1.0f } } } };
+#endif
```
Why did you hardcode disable this code with directives? Should it be removed?

shinyquagsire23 replied:
could remove it yeah, wasn't sure if I'd have to come back to this code or not

zmerp commented on alvr/server_openvr/cpp/platform/win32/OvrDirectModeComponent.h:
```diff
@@ -71,5 +78,8 @@ class OvrDirectModeComponent : public vr::IVRDriverDirectModeComponent {
     uint64_t m_targetTimestampNs;
     uint64_t m_prevTargetTimestampNs;
 
+    vr::HmdRect2_t m_viewProj[2];
+    vr::HmdMatrix34_t m_eyeToHead[2];
```
I'm not convinced about saving the view parameters in the DirectMode class. I think they should stay inside `m_pD3DRender`

zmerp commented on alvr/server_openvr/cpp/platform/win32/CEncoder.cpp:
```diff
@@ -91,7 +94,9 @@ bool CEncoder::CopyToStaging(
     m_FrameRender->Startup();
 
-    m_FrameRender->RenderFrame(pTexture, bounds, layerCount, recentering, message, debugText);
+    m_FrameRender->RenderFrame(
+        pTexture, bounds, poses, viewProj, eyeToHead, layerCount, recentering, message, debugText
+    );
```
I'm also not convinced why the renderer call should be forwarded from the encoder. It doesn't make sense that the renderer is "owned" by the encoder. I don't think this was like that initially but then the linux implementation was made like that and so also the windows impl got changed.
Don't bother fixing this in this PR.

zmerp reviewed (commented):
Sorry it took a bit, left some comments.

shinyquagsire23 commented:
@zmerp  Addressed the nits, I guess I kinda see the logic on `m_pD3DRender` since it would allow render passes to do things with per-eye transforms and such if they needed to w/o a bunch of arg passing, but it still seems a bit odd to me.

zmerp reviewed (approved):
Looks nice now. I didn't test but either this code is all correct it or it will crash, edge cases are unlikely. Your testing should suffice.

zmerp reviewed (changes requested):
Oh wait, my bad. I said the class d3dRender but I meant FrameRender. I thought they were the same class.
